Exposed toggle scrolling as a key in via

beaudrychase opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

beaudrychase commented on keyboards/ploopyco/madromys/madromys.h:
```diff
@@ -33,5 +33,5 @@ extern uint16_t          dpi_array[];
 enum ploopy_keycodes {
     DPI_CONFIG = QK_KB_0,
-    DRAG_SCROLL,
+    DRAG_SCROLL, TOGGLE_SCROLL
```
This will let people define the behavior for scrolling with VIA. They could even have both keys if they want. I wasn't able to get TOGGLE_SCROLL to be accepted in the UI but you can edit the layout.json to have `"CUSTOM(2)"` for a key to make it toggle scroll.

drashna commented:
Just a heads up, develop has a pretty massive rework of all of the ploopy code, to unify it into a more cohesive base.   As such, the files that you're editing no longer exist on that branch, and this change will cause merge conflicts.

I would recommend rebasing this to target develop instead.

beaudrychase commented:
Oh sick, I hadn't seem that PR. Honestly, this was a quick and dirty PR for me to get the functionality I wanted and something to link to in [this issue](https://github.com/ploopyco/adept-trackball/issues/9#issuecomment-2115312506). I can for sure clean this up and rebase on top of that commit you linked.

beaudrychase commented on keyboards/ploopyco/ploopyco.c:
```diff
@@ -41,6 +41,9 @@
 #    endif
+#ifndef PLOOPY_DRAGSCROLL_MOMENTARY
+#    define PLOOPY_DRAGSCROLL_MOMENTARY 1
+#endif
```
On a completely separate note from this PR. https://github.com/qmk/qmk_firmware/pull/22967 these changes changed how the sensitivity for scrolling is defined. It seems that the default is really sensitive.

beaudrychase replied:
https://github.com/beaudrychase/qmk_firmware/blob/38d6e9a75a4cff4ae5bd32209fc0cecb83c6f23b/keyboards/ploopyco/madromys/madromys.c#L32 The old way scroll sensitivity was configured was through changing the dpi settings while drag scroll was enabled. The default dpi for Adept was 900 and the drag scroll dpi was 100. This means that if the `PLOOPY_DRAGSCROLL_DIVISOR` was set to 9 instead of 8, the sensitivity should be the same for most people compared to their old firmware.

drashna replied:
If anything, if we're adding a separate dragscroll keycode, then I'd opt for completely removing this define.  So just one keycode for momentary, and one for toggle, and leave that behavior hard coded.  It makes things a bit less confusing for users.

beaudrychase commented on keyboards/ploopyco/ploopyco.c:
```diff
@@ -127,30 +130,26 @@ void encoder_driver_task(void) {
+        mouse_report.h = (mouse_report.x + scroll_remainder_h) / PLOOPY_DRAGSCROLL_DIVISOR_H;
 #ifdef PLOOPY_DRAGSCROLL_INVERT
-        mouse_report.v = -(int8_t)scroll_accumulated_v;
+        mouse_report.v = -(mouse_report.y + scroll_remainder_v)  / PLOOPY_DRAGSCROLL_DIVISOR_V;
```
This method avoids conversions to and from floats which can introduce inaccuracies when values are accumulated and integer math is faster.

drashna replied:
Less float math is better, especially for the atmega32u4 based boards!

beaudrychase commented on keyboards/ploopyco/ploopyco.c:
```diff
@@ -127,30 +130,26 @@ void encoder_driver_task(void) {
-
-        mouse_report.x = 0;
-        mouse_report.y = 0;
+    } else {
+        scroll_remainder_h = 0;
+        scroll_remainder_v = 0;
```
This is so the remainders from a previous drag scroll use aren't at the start of a new one.

beaudrychase commented on keyboards/ploopyco/readme.md:
```diff
@@ -33,13 +33,14 @@ The `DPI_CONFIG` macro will cycle through the values in the array, each time you
 ### Drag Scroll Configuration
 
-* `#define PLOOPY_DRAGSCROLL_MOMENTARY` - Makes the key into a momentary key, rather than a toggle.
+* `#define PLOOPY_DRAGSCROLL_MOMENTARY` - Makes `DRAG_SCROLL` into a momentary key, rather than a toggle.
```
I thought it best to preserve the behavior of this macro even though this is explicitly the use case for `TOGGLE_DRAG_SCROLL`. I wasn't sure what would be preferred, to preserve it or just delete the macro. Preserving it felt more conservative.

beaudrychase commented on keyboards/ploopyco/ploopyco.h:
```diff
@@ -33,6 +33,7 @@ extern uint16_t          dpi_array[];
 enum ploopy_keycodes {
     DPI_CONFIG = QK_KB_0,
     DRAG_SCROLL,
+    TOGGLE_DRAG_SCROLL
```
This will let people define the behavior for scrolling with VIA. They could even have both keys if they want. I wasn't able to get TOGGLE_SCROLL to be accepted in the UI but you can edit the layout.json to have "CUSTOM(2)" for a key to make it toggle scroll.

beaudrychase commented on keyboards/ploopyco/ploopyco.c:
```diff
@@ -127,30 +127,25 @@ void encoder_driver_task(void) {
-        mouse_report.v = -(int8_t)scroll_accumulated_v;
-#else
-        mouse_report.v = (int8_t)scroll_accumulated_v;
+        mouse_report.v *= -1;
```
Should this macro be called `PLOOPY_DRAGSCROLL_INVERT_V`?

drashna replied:
Makes sense to me.  Makes it much clearer what it is for.

beaudrychase commented:
This is my first time contributing to this project; is there anything I need to do to get this merged in? @drashna I also think this PR is in a done state. @drashna thank you so much for your work on the Ploopy firmware and for taking the time to review this PR.

drashna commented:
Generally, we want two approvals from collaborators or other exports for the PR before it's merged.

So, it's mostly a waiting game. And usually, it can take a couple of weeks.  There are a lot of PRs, and we all do this in our free time.

beaudrychase commented:
That's a sensible policy. Thank you again for all of your efforts!

beaudrychase commented on keyboards/ploopyco/ploopyco.c:
```diff
@@ -175,7 +175,7 @@ bool process_record_kb(uint16_t keycode, keyrecord_t* record) {
 
     if (keycode == MOMENTARY_DRAG_SCROLL) {
-        is_drag_scroll = record->event.pressed;
+        is_drag_scroll ^= 1
```
This was a subtle thing I wanted to change.
Behavior before this commit:
1. Press TOGGLE_DRAG_SCROLL and enter drag scroll mode.
2. Press and hold MOMENTARY_DRAG_SCROLL leaves drag scroll mode and enters mouse mode.
3. Release MOMENTARY_DRAG_SCROLL does nothing, the device stays in mouse mode instead of going back into drag scroll.

Behavior after this commit:
1. Press TOGGLE_DRAG_SCROLL and enter drag scroll mode.
2. Press and hold MOMENTARY_DRAG_SCROLL temporarily leaves drag scroll mode and enters mouse mode.
3. Release MOMENTARY_DRAG_SCROLL returns device to drag scroll mode
4. Press TOGGLE_DRAG_SCROLL to leave drag scroll mode.

beaudrychase commented:
I just read this PR: https://github.com/qmk/qmk_firmware/pull/21426
It seems like my work here won't be needed once that one is merged in since ploopyco won't have to implement drag scroll. That PR also exposes momentary and toggle versions of drag scroll as keycodes. These ploopyco will just need to define `POITNING_DEVICE_MODES_ENABLE`. I'd be cool with just closing this PR, it's up to y'all.

beaudrychase commented on keyboards/ploopyco/madromys/config.h:
```diff
@@ -18,6 +18,8 @@
 
 #pragma once
 
+#define VIA_FIRMWARE_VERSION 1
```
@light-idea Thanks for your suggestion to version bump VIA. Does this have the intended effect? I followed the commit you linked in your comment.

I was also wondering about your choice of `config.h`. You chose to modify `.../ploopyco/trackball/rev1_005/config.h` instead of `.../ploopyco/trackball/config.h`. I think defining it in either should have the intended effect. What considerations are there with this?

beaudrychase commented on keyboards/ploopyco/ploopyco.h:
```diff
@@ -32,7 +32,8 @@ extern uint16_t          dpi_array[];
 enum ploopy_keycodes {
     DPI_CONFIG = QK_KB_0,
-    DRAG_SCROLL,
+    TOGGLE_DRAG_SCROLL,
+    MOMENTARY_DRAG_SCROLL,
```
@light-idea good catch with the ordering of these.

bradfordwagner commented:
@beaudrychase  - Thanks for your contributions. I want to encourage you to continue pushing to get this PR merged. I can't find another way of enabling TOGGLE_DRAG_SCROLL with custom DPI configurations, the ploopy appears to only have pre-compiled versions of this custom keybind. Your fork/branch works perfectly

beaudrychase commented:
Changes since last review:
1. Rebased to the top of origin/develop.
2. Removed via references
3. Backwards compatibility with `PLOOPY_DRAGSCROLL_MOMENTARY` directive. It only impacts the original keycode.

beaudrychase commented:
Changes since last review:
1. Rebased to the top of origin/develop.
2. Removed via references
